ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Mempool.hs

```diff
@@ -96,18 +98,91 @@ instance ( CanHardFork xs
          , HasCanonicalTxIn xs
          , NoThunks (Ticked1 (LedgerState (HardForkBlock xs)) TrackingMK)
          ) => LedgerSupportsMempool (HardForkBlock xs) where
-  applyTx   = applyHelper ModeApply
+  applyTx = applyHelper ModeApply
 
-  reapplyTx = \cfg slot vtx tls ->
-        fmap (\(tls', _vtx) -> tls')
-      $ applyHelper
+  reapplyTx cfg slot vtx tls =
+          fst
+      <$> applyHelper
           ModeReapply
           cfg
           DoNotIntervene
           slot
           (WrapValidatedGenTx vtx)
           tls
 
+  reapplyTxs
+    HardForkLedgerConfig{..}
+    slot
+    vtxs
+    (TickedHardForkLedgerState transition hardForkState) =
+      (\(err, val, st') ->
+           (mismatched' ++ err, val, TickedHardForkLedgerState transition $ State.HardForkState st'))
+      . mySequence
+      . State.getHardForkState
+      $ hcizipWith proxySingle modeApplyCurrent cfgs matched
+
+    where
+      pcfgs = getPerEraLedgerConfig hardForkLedgerConfigPerEra
+      cfgs  = hcmap proxySingle (completeLedgerConfig'' ei) pcfgs
+      ei    = State.epochInfoPrecomputedTransitionInfo
+                hardForkLedgerConfigShape
+                transition
+                hardForkState
+
+      -- Transactions are unwrapped into the particular era transactions.
+      (mismatched, matched) =
+        matchPolyTxs
+          -- How to translate txs to later eras
+          (InPairs.hmap product2Snd (InPairs.requiringBoth cfgs hardForkInjectTxs))
+          (map (getOneEraValidatedGenTx . getHardForkValidatedGenTx) vtxs)
+          hardForkState
+
+      mismatched' :: [Invalidated (HardForkBlock xs)]
+      mismatched' =
+        map (Invalidated . HardForkValidatedGenTx . OneEraValidatedGenTx . fst) mismatched
+
+      modeApplyCurrent :: forall blk.
+           SingleEraBlock                                blk
+        => Index xs                                      blk
+        -> WrapLedgerConfig                              blk
+        -> Product
+             ([] :.: ((,) (NS WrapValidatedGenTx xs) :.: WrapValidatedGenTx))
+             (FlipTickedLedgerState ValuesMK)            blk
+        -> (  (,,)
+              [Invalidated (HardForkBlock xs)]
+              [Validated (GenTx (HardForkBlock xs))]
+            :.:
+              FlipTickedLedgerState DiffMK)                blk
+      modeApplyCurrent index cfg (Pair (Comp txs) (FlipTickedLedgerState st)) =
+        let (err, val, st') =
+              reapplyTxs (unwrapLedgerConfig cfg) slot [ unwrapValidatedGenTx t | Comp (_, t) <- txs ] st
+        in Comp
+           ( map (Invalidated . injectValidatedGenTx index . getInvalidated) err
+           , map (HardForkValidatedGenTx . OneEraValidatedGenTx . injectNS index . WrapValidatedGenTx) val
+           , FlipTickedLedgerState st'
+           )
+
+      mySequence ::
```

Comment:
Note to reviewer: if there is something that implements this in SOP I would like to use that instead of this custom one, but I don't think there is

Reply:
Thanks!!!!!